raft: Pipeline hints into MsgVoteResp

hakuuww opened the pull request:
This PR optimizes the first MsgApp/Replication flow after a raft candidate becomes the new leader. 

**Commit by commit review recommended.**

This is done by Pipelining/piggybacking a hint into `MsgVoteResp` and having the candidate use that hint:
1. The voter attaches a best guess hint `entryID` into `MsgVoteResp` when reacting to `MsgVote`. 
2. Upon receiving `MsgVoteResp`, the candidate computes a `matchGuess` index if `MsgVoteResp` is non-reject.
3. The candidate keeps those conflict `entryIDs` in memory, and send `MsgApp` starting from there after winning election. 
4. If the voter voted for this candidate, and the voter's `hint EntryID` "suggests a match" on a leader's log, the leader transfers `Progress.State` to `StateReplicate` for that follower when becoming the leader after winning election.

Subsequent `MsgApp` to the followers that are already in `StateReplicate` are "(almost) guaranteed" to succeed and carries more raft entries than in `StateProbe`. And so the leader can commit faster (assuming moderate incoming raft proposals).

Even for followers (voters) where the candidate received a vote with hint from, but could not set to `StateReplicate` (because the hint does not suggest a match), the first round of `MsgApp` is still very likely to succeed. This is because we computed a best guess `next Index` based on the original hint(which was also a best guess on where the two logs match).

Both the voter and candidate side computes a (regressed) best guess (index,term) using findConflictByTerm, this is very similar to how a leader and a follower resolve log conflicts. 

If the hint suggests that the voter would need a snapshot to be caught up, a snapshot would be sent by if the candidate wins election.

---
**Why this PR brings some performance improvements:** 

In the happy scenario, all the supporting voters attaches a `lastEntryID` hint, the candidate computes a `conclictIndex`, which is very likely to be a match. It would only not be a match if we have a long divergent tail with conflicting terms (possible, but very unlikely to happen).

So when the candidate wins election and becomes leader, it can set a quorum of followers to `StateReplicate`. 
This allows for faster replication and commit time. Essentially saving "1 RTT" in "almost all"(like 95%+?)  leader election scenarios. 

---

**Why we can do this with a high level informal proof:**
Due to the following guarantees by Raft: 
(deduced from Raft paper section 5.2 & 5.4)

  - When becoming leader from candidate state, the candidate/leader does not overwrite any of its own log, it simply appends a dummy entry with a new term number when becoming leader.

  - If a voter "`B`" voted for candidate `A`, then it means voter `B` won't vote for anybody else for the current election term(of candidate `A`). Also, voter `B` won't accept any `MsgApp` from nodes of lower or equal term than `A`.

NB: `B`'s log can be overwritten, by a different leader `C` with term higher than `A`, which is fine, since `A` will also get overwritten by that different leader `C`.

---

**Future optimizations 1:**
A candidate becomes the leader immediately when receiving enough votes(quorum), so the `MsgVoteResp` from the remaining replica(s) are ignored in this case. 

This is largely fine though, since we only need a quorum to make progress(commit new entries). The other replica can lag behind by a bit(behind by just 1rtt in most cases). 

But that hint would be useful, if it allows for catching up a candidate faster.
 
The change would involve letting the leader handle `MsgVoteResp` in a similar way as `MsgAppResp`.

---

**Further optimizations 2:**

In this issue, we are only attaching a single hint index which is the voter's last entryID in `MsgVoteResp`. 

In extreme log inconsistency scenarios, a single hint index would not satisfy as a solution. 

Example:
Imagine the following case, which is a somewhat extreme example, but is theoretically possible. 

```
#      1  2  3  4  5  6  7  8  9 10   index
# n1: [1][1][3][3][5][5][7]
# n2: [1][1][2][2][4][4]
```

n1 is the new leader of term 7 with dummy entry: t7/8 
n2 is a particular voter

1. n1 sends `MsgVote` to n2, with its `candLastID` = (index:8, term:7)
2. n2 votes for n1, computes a hint entryID =`(index:6, term:4)` based on `candLastID` = `(index:7, term:7)`, sends `MsgVoteResp` to n1
3. n1 receives `MsgVoteResp` from n2, compute a `matchGuess` index  based on the hint `(index:6, term:4)`
  the `matchGuess` is at (index: 4, term:3). This does not suggest a match with the voter's log.
4. n1 wins election after receiving a quorum of votes(including the vote from n2). Sends out the first round of MsgApp starting at (index: 5), which would not succeed. 

Overall, An extra round of probing must happen before n1 can replicate to n2 in steady flow.

However, in `MsgVoteResp`, we can actually attach the suffix entryIDs on the voter's raftLog which is the `termCache`. 
If we do that, then we have the guarantee(99.9% chance) that we can (almost) always transfer a quorum of followers into `stateReplicate` when becoming leader. This guarantee will be quite useful for upper layers that rely on Raft.

For this example, if n2 attaches the term cache into 'MsgVoteResp{reject = false}', then the leader can compute the correct match index t1/2 from the term cache, set n2's state to `stateReplicate`, and start replicating at t3/3 which will succeed. 

---


References: https://github.com/cockroachdb/cockroach/issues/142492
Epic: None
Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/143713)
<!-- Reviewable:end -->

hakuuww commented on pkg/raft/testdata/probe_and_replicate.txt:
```diff
@@ -787,21 +747,38 @@ stabilize 1 6
+  Committed: (18,21]
+  Messages:
+  6->1 MsgAppResp Term:8 Log:0/21 Commit:21
+  Applying:
```
about this part, has to do with leader fortification.

pav-kv commented on pkg/raft/tracker/electiontracker.go:
```diff
@@ -16,21 +16,24 @@ type ElectionTracker struct {
 	config *quorum.Config
 
 	votes map[pb.PeerID]bool
+	hints map[pb.PeerID]pb.Entry
```
I would structure this as:

```golang
type vote struct {
	acc  bool    // vote accepted if true, rejected if false
	last entryID // last entry ID in the voter's log
}

votes map[pb.PeerID]vote
```

So that we don't have to allocate yet another map.

pav-kv commented on pkg/raft/raft.go:
```diff
@@ -1066,12 +1066,18 @@ func (r *raft) reset(term uint64) {
 
-	r.electionTracker.ResetVotes()
 	r.trk.Visit(func(id pb.PeerID, pr *tracker.Progress) {
+		guessMatch := r.raftLog.lastIndex()
+		hintEntry, ok := r.electionTracker.GetHintEntry(id)
+		conflictIndex, t := r.raftLog.findConflictByTerm(hintEntry.Index, hintEntry.Term)
+		if ok && t != 0 {
+			guessMatch = conflictIndex
+		}
```
This part shouldn't be in every `raft.reset`. We only need to do it when `becomeLeader`.

pav-kv commented on pkg/raft/raft.go:
```diff
@@ -1066,12 +1066,18 @@ func (r *raft) reset(term uint64) {
+			guessMatch = conflictIndex
+		}
+
 		*pr = tracker.Progress{
```
This still initializes all flows to `StateProbe`, but the point of all this was also to initialize to `StateReplicate` in best case.

The two modifications can be done in separate commits though:
- in this commit we improve the initial guess (by setting `Next`)
- in another commit we can also enter `StateReplicate` for those flows for which we have a match

pav-kv reviewed (commented):
> Is there a better approach to do this? I don't know if it is clean to put it here in the election tracker. But now we mash hints into this tracker which is a bit unclean.
...
> A reasonable alternative is to compute the conflict index on the spot when receiving MsgVoteResp, and storing the conflictIndex, instead of computing it when becoming leader.

Seems like, either way, you need to store either an unprocessed or post-processed hint somewhere. I would argue that saving the unprocessed hint in election voter is clean enough. Also, don't call it "hint" - make it `last entryID` since that's what it is now.

Alternatively, could initialize `Progress` map when in `StateCandidate` and apply the hints to it in some way, instead of saving them to election tracker.

hakuuww committed:
raft: log hint EntryID explicitly if it is non-zero

Just better overall readability for (reject) hints entryIDs.

Epic: none
Release note: None

pav-kv commented on pkg/raft/raft.go:
```diff
@@ -1329,7 +1338,46 @@ func (r *raft) becomeLeader() {
+		}
+	}
+
+	r.reset(r.Term, specialVisit)
```
Perhaps the `r.trk.Visit` below can be merged into this visitor. The `TODO(pav-kv)` a few lines below already encompasses this, so let's do in a follow-up.

hakuuww committed:
raft: refactor reset() to pass in special visitor

Previously, the r.reset() has its own internal in-line visit logic to reset all
the peers to a default state through r.trk.Visit().

Now, a new function variable parameter is introduced to reset(), which allows
the caller of r.reset() to pass in a special function which goes into
r.trk.Visit().

Callers of reset() are updated.

Epic: none
Release note: None

hakuuww commented on pkg/raft/raft.go:
```diff
@@ -1337,7 +1338,46 @@ func (r *raft) becomeLeader() {
+
+	// A special visit function to use the recorded hints and optimize the first
+	// round of MsgApp after becoming leader.
+	specialVisit := func(id pb.PeerID, pr *tracker.Progress) {
```
Done

pav-kv commented on pkg/raft/tracker/electiontracker.go:
```diff
@@ -12,25 +12,43 @@ import (
-	votes map[pb.PeerID]bool
+// hintInfo stores information needed for the new leader to decide where to send
+// entryIDs from after winning leader election.
+type hintInfo struct {
+	// conflictIdx is the best guess on where the new leader should send to the
+	// follower if the candidate wins the election.
+	conflictIdx pb.Index
```
Naming nit.

```golang
// matchGuess stores ...
type matchGuess struct {
	// index is the best guess on where the voter's log potentially matches the
	// candidate's. When elected, the leader will start replicating the log from
	// index + 1.
	index pb.Index
	// match indicates whether the voter's log matches the candidate's at index.
	match bool
}
```

hakuuww replied:
Done

pav-kv commented on pkg/raft/tracker/electiontracker.go:
```diff
@@ -12,25 +12,43 @@ import (
-		votes:  map[pb.PeerID]bool{},
+		config:       config,
+		votes:        map[pb.PeerID]bool{},
+		conflictInfo: map[pb.PeerID]hintInfo{},
```
Try avoiding the naming/terminology explosion. Here we have "conflict", "hint", "guess", etc. We can squash it all:

```golang
matchGuess: map[pb.PeerID]matchGuess
```

hakuuww replied:
Would it be confusing to have the same name for map, and map's content?

pav-kv replied:
Looks fine to me. Plural like `matchGuesses` / `matchHints` is also acceptable, no strong opinion here. Important bit was to reduce terminology.

pav-kv commented on pkg/raft/tracker/electiontracker.go:
```diff
@@ -55,9 +73,28 @@ func (e *ElectionTracker) TallyVotes() (granted int, rejected int, _ quorum.Vote
+// on where to send to the follower or just r.raftLog.LastIndex().
+func (e *ElectionTracker) GetGuessMatchEntryID(
+	id pb.PeerID,
+) (conflictID pb.Index, match bool, exist bool) {
```
`conflictID` isn't a right name since it's not an ID. Also, how about returning `MatchGuess` instead of index/bool?

hakuuww replied:
Removed

pav-kv commented on pkg/raft/tracker/electiontracker.go:
```diff
@@ -55,9 +73,28 @@ func (e *ElectionTracker) TallyVotes() (granted int, rejected int, _ quorum.Vote
+// follower's log matches the entryID on the leader's log.
+// If match is false, it means the entryID at conflictIdx is either a best guess
+// on where to send to the follower or just r.raftLog.LastIndex().
+func (e *ElectionTracker) GetGuessMatchEntryID(
```
In Go, `Get` is usually a redundant name: https://google.github.io/styleguide/go/decisions#getters

Consider `func MatchGuess(id pb.PeerID) (MatchGuess, bool) { ... }`.

hakuuww replied:
For GetVote() (I removed it), would it be good practice to use GetVote()? since vote might be more confusing, its a noun and verb at the same time

pav-kv replied:
Good point. Yeah, when ambiguous, adding `Get` makes things clearer.

hakuuww commented on pkg/raft/raft.go:
```diff
@@ -2357,6 +2377,56 @@ func leadSliceFromMsgApp(m *pb.Message) LeadSlice {
+	match := false
+	// Compute a conflictIndex which is a best guess on where the voter's log
+	// matches the leader's.
+	conflictIndex, conflictTerm := r.raftLog.findConflictByTerm(voterLastEntryID.index, voterLastEntryID.term)
```
```
#      1  2  3  4  5  6  7  8  9 10   index
# n1: [1][1][4][4]
# n2: [1][1][3][3][4]
```
n1 is the candidate, n2 is the voter
Can this scenario occur?
Voter attaches t4/5 as hint, the candidate calls findConflictByTerm, returns index = 5, term = 4

this evaluates to a match, but it is in fact not a match. 

I think this is not possible to occur, due to raft invariants.

hakuuww committed:
raft: take out logic for candidate handling vote response

This logic will become quite lengthy, move to seperate function.

Epic: None
Release Note: None

hakuuww commented on pkg/raft/raft.go:
```diff
@@ -2332,6 +2384,57 @@ func leadSliceFromMsgApp(m *pb.Message) LeadSlice {
+	// this candidate wins election, the voter's log will not change until the
+	// first MsgApp from this or later leader.
+	// (See section 5.2 & 5.4 of the Raft paper for more details.)
+	if myVoteRespType == pb.MsgVoteResp && !m.Reject && m.LogTerm > 0 {
```
`if myVoteRespType == pb.MsgVoteResp && !m.Reject && m.LogTerm > 0 && m.RejectHint > 0 {`
It may be safer to include check for m.RejectHint > 0
Just in case LogTerm is set to non-zero accidentally, it doesn't happen today

hakuuww commented on pkg/raft/tracker/electiontracker.go:
```diff
@@ -12,25 +12,43 @@ import (
-// instance if v == true (and declined it otherwise).
-func (e *ElectionTracker) RecordVote(id pb.PeerID, vote bool) {
+// instance if vote == true (and declined it otherwise).
+func (e *ElectionTracker) RecordVote(id pb.PeerID, vote bool, conflictIdx pb.Index, match bool) {
```
I exported the type, the reason I broke it down into index/bool is to separate the election tracker with raft. Is having MatchGuess exported acceptable here?

pav-kv replied:
Yeah, looks better. It's a good practice to group related things into a type. A type carries semantics. E.g. the reader can jump to the definition of the type and read what it means. Also, stronger typing often removes footguns.

pav-kv commented on pkg/raft/raft.go:
```diff
@@ -1033,7 +1033,7 @@ func (r *raft) maybeCommit() bool {
 }
 
-func (r *raft) reset(term uint64) {
+func (r *raft) reset(term uint64, specialVisit func(id pb.PeerID, pr *tracker.Progress)) {
```
Yeah, nil is better. Variadics allocate a slice under the hood (unless Go compiler optimizes it), so better stay away from it if not needed.

pav-kv commented on pkg/raft/tracker/electiontracker.go:
```diff
@@ -55,9 +70,15 @@ func (e *ElectionTracker) TallyVotes() (granted int, rejected int, _ quorum.Vote
 	return granted, rejected, result
 }
 
+func (e *ElectionTracker) GuessMatch(id pb.PeerID) (MatchGuess, bool) {
```
Could you add a comment for this method? All exported methods should have a comment.

nit: name it `MatchGuess` to match what it returns?

hakuuww replied:
Done

pav-kv commented on pkg/raft/raft.go:
```diff
@@ -1342,7 +1342,41 @@ func (r *raft) becomeLeader() {
+	// A special visit function to use the recorded hints and optimize the first
+	// round of MsgApp after becoming leader.
+	r.reset(r.Term, func(id pb.PeerID, pr *tracker.Progress) {
+		// next is set to r.raftLog.lastIndex() + 1 by default like in reset().
+		next := r.raftLog.lastIndex() + 1
+		// ConflictIdx is either a best guess on where the voter's log matches the
+		// leader's, or r.raftLog.lastIndex()
+		matchGuess, ok := r.electionTracker.GuessMatch(id)
+		conflictIdx := matchGuess.Index
+		isMatch := matchGuess.Match
+		// A conflictIdx is recorded and the voter voted during election.
+		if ok {
+			next = uint64(conflictIdx) + 1
+		}
```
Most of these variables are not needed.

```golang
// ...
matchGuess := tracker.MatchGuess{Index: r.raftLog.lastIndex()}
if guess, ok := r.electionTracker.MatchGuess(id); ok {
	matchGuess = guess
}

*pr = tracker.Progress {
	...
	Next: 1, // invariant: Next > Match
	...
}

if matchGuess.Match {
	becomeReplicate()
}
// Comment
pr.Next = matchGuess.Index + 1
```

hakuuww replied:
Done

pav-kv commented on pkg/raft/testdata/lazy_replication.txt:
```diff
@@ -48,15 +48,16 @@ stabilize
 # Attempt to send a misaligned MsgApp. No-op.
 send-msg-app 1 to=2 lo=10 hi=13
 ----
-could not send MsgApp (10,13] to 2
```
Could you update the `# comments` above this command and the one below it? Now, the first send is aligned, and the second one isn't. This looks fine.

pav-kv commented on pkg/raft/testdata/probe_and_replicate.txt:
```diff
@@ -555,77 +555,67 @@ stabilize 1
+----
+1: StateReplicate match=21 next=22 sentCommit=18 matchCommit=18
+2: StateReplicate match=0 next=22 sentCommit=18 matchCommit=0 inactive inflight=1
+3: StateReplicate match=0 next=22 sentCommit=18 matchCommit=0 inactive inflight=1
```
Could you put a `# comment` above this command, explaining what we're seeing here?

pav-kv commented on pkg/raft/testdata/snapshot_succeed_via_app_resp_behind.txt:
```diff
@@ -120,23 +116,12 @@ INFO 3 [commit: 11] restored snapshot [index: 11, term: 1]
-> 1 handling Ready
-  Ready:
-  Messages:
-  1->3 MsgSnap Term:1 Log:0/0
```
Looks like this test now doesn't do what it intended. We need to coerce it into sending this snapshot again, and also update the comments above the commands.

pav-kv replied:
SGTM, the fix seems minimal.

hakuuww committed:
raft: voter attaches a hint entryID in MsgVoteResp message

Similar to how Reject hint entryID is populated in MsgAppResp{Reject = true},

When a voter votes for a candidate with MsgVoteResp{Reject = false},
message.RejectHint and message.LogTerm are populated with a best guess hint
entryID based on candidate's lastID.

Epic: None
Release note: None

hakuuww committed:
raft: print follower states after becoming leader in probe_and_replicate.txt

Also stabilize n7 in the end.
(This was previously removed for whatever reason, it is good to keep it in)

Epic: none
Release note: None

hakuuww committed:
raft: print follower states after becoming leader in lazy_replication.txt

Epic: none
Release note: None

hakuuww reviewed (commented):
Addressed

hakuuww commented on pkg/raft/raft.go:
```diff
@@ -2347,8 +2347,40 @@ func leadSliceFromMsgApp(m *pb.Message) LeadSlice {
-	gr, rj, res := r.poll(m.From, m.Type, !m.Reject)
+func (r *raft) handleVoteResp(m pb.Message, myVoteRespType pb.MessageType) {
+	matchGuess := r.raftLog.lastIndex()
+	match := false
```
done

hakuuww commented on pkg/raft/raft.go:
```diff
@@ -1744,7 +1744,24 @@ func (r *raft) Step(m pb.Message) error {
 			// The term in the original message and current local term are the
 			// same in the case of regular votes, but different for pre-votes.
-			r.send(pb.Message{To: m.From, Term: m.Term, Type: voteRespMsgType(m.Type)})
+			var logTerm, rejectHint uint64
```
done

hakuuww commented on pkg/raft/raft.go:
```diff
@@ -1744,7 +1744,24 @@ func (r *raft) Step(m pb.Message) error {
 			// same in the case of regular votes, but different for pre-votes.
-			r.send(pb.Message{To: m.From, Term: m.Term, Type: voteRespMsgType(m.Type)})
+			var logTerm, rejectHint uint64
+			if voteRespMsgType(m.Type) == pb.MsgVoteResp {
```
done

pav-kv commented on pkg/raft/raft.go:
```diff
@@ -1342,7 +1342,38 @@ func (r *raft) becomeLeader() {
+			pr.Next = matchGuess.Index + 1
+		}
+
+		if pr.Next <= r.raftLog.storage.Compacted() {
+			pr.RecentActive = true
+		}
```
As discussed in Slack, we can make this less corner-casey by saying that all peers with a response become `RecentActive`. This is consistent with the logic of setting `RecentActive` to true on all `Resp` messages.

E.g. a simple fix:

```golang
matchGuess := tracker.MatchGuess{Index: r.raftLog.lastIndex()}
replied := false
if guess, ok := r.electionTracker.MatchGuess(id); ok {
	matchGuess, replied = guess, true
}

*pr = Progress {
	...
	RecentActive: replied,
}
```

pav-kv commented on pkg/raft/raft_test.go:
```diff
@@ -2685,6 +2685,194 @@ func TestLeaderAppResp(t *testing.T) {
+			// Create candidate node
+			r := newTestRaft(1, 5, 1, storage)
+			r.Term = tc.candidateLog.term
+			r.Step(pb.Message{From: 1, To: 1, Type: pb.MsgHup})
+			r.advanceMessagesAfterAppend()
```
Probably simpler to say `r.becomeCandidate()`, other tests do that too.

pav-kv commented on pkg/raft/raft_test.go:
```diff
@@ -2685,6 +2685,194 @@ func TestLeaderAppResp(t *testing.T) {
+			// Check that node 2's Progress.Next is set based on the best guess
+			p := r.trk.Progress(2)
+			msgs := r.readMessages()
+			called := 0
+			for _, msg := range msgs {
+				// Because the Progress.Next on the leader side will be updated when the
+				// leader sends out the first round of MsgApp.
+				// Extract the pending messages to check for the Progress.Next that was
+				// set when the candidate became leader.
+				if (msg.Type == pb.MsgSnap || msg.Type == pb.MsgApp) && msg.To == 2 {
+					assert.Equal(t, tc.next, msg.Index+1)
+					called += 1
+				}
+			}
```
I think I got the idea, but perhaps it would be less cumbersome to set `LazyReplication` flag to true. Then `Next` (I think) won't move forward.

pav-kv commented on pkg/raft/raft.go:
```diff
@@ -2355,11 +2394,98 @@ func (r *raft) handleVoteResp(m pb.Message, myVoteRespType pb.MessageType) {
+		// match is wrongly set to true because term == voterLast.term
+		// But this is NOT POSSIBLE, because the candidate's log at index 7 is
+		// because the 2 raft logs violates Raft invariants.
+		index, term := r.raftLog.findConflictByTerm(min(voterLast.index, r.raftLog.lastIndex()), voterLast.term)
```
Initially I thought this is incorrect, but after some contemplation it makes sense.

Although I think we need to change the semantics of `findConflictByTerm` method (at least in its comment), because it currently assumes that the passed in parameters are a real entry ID. Whereas now it accepts an "upper bound" of a term at a specific index.

Wondering if this `min` could move inside `findConflictByTerm`. The contract would become that `findConflictByTerm` finds an existing index (i.e. <= `lastIndex`) such that its term <= the given one. And that would be nice: minus one corner case in it.

A note about the big comment above: it can overwhelm / scare the reader. There is probably a shorter way to express it, and for specific examples it can reference the test (e.g. say `see TestVoteHintPipelining for concrete examples`).

pav-kv replied:
I think we want this: #145192.

hakuuww replied:
> > I think we want this: https://github.com/cockroachdb/cockroach/pull/145192.
> 
> Another confirmation that this optimization should work is that we already do this on a follower's end:
> 
> [cockroach/pkg/raft/raft.go](https://github.com/cockroachdb/cockroach/blob/c7a69de62b16bd0920459ee0c202cec46d013d61/pkg/raft/raft.go#L2398-L2399)
> 
> Lines 2398 to 2399 in [c7a69de](https://github.com/cockroachdb/cockroach/commit/c7a69de62b16bd0920459ee0c202cec46d013d61)
> 
>  hintIndex := min(m.Index, r.raftLog.lastIndex()) 
>  hintIndex, hintTerm := r.raftLog.findConflictByTerm(hintIndex, m.LogTerm) 
> So doing it symmetrically at the leader's end seems appropriate:
> 
> [cockroach/pkg/raft/raft.go](https://github.com/cockroachdb/cockroach/blob/c7a69de62b16bd0920459ee0c202cec46d013d61/pkg/raft/raft.go#L2020)
> 
> Line 2020 in [c7a69de](https://github.com/cockroachdb/cockroach/commit/c7a69de62b16bd0920459ee0c202cec46d013d61)
> 
>  nextProbeIdx, _ = r.raftLog.findConflictByTerm(m.RejectHint, m.LogTerm) 
> 
> Although it will be a no-op change in this second case because a follower can't reply with a hint above the leader's lastIndex in this case.

> Yes, turns out this optimization is not that involved after all, further improvement would involve attaching the candidate lastEntryID into MsgVote, and have the voter going through findConflictByTerm based on the leader lastEntryID.

In that case the candidate will not receive a hint higher than its own log in MsgVoteResp.

pav-kv reviewed (commented):
Looks good. I'll make another pass since I'd like to contemplate more on the logic in `handleVoteResp`. It sounds good but maybe needs some proof. Feel free to ignore nits, I might just push a commit fixing them.

pav-kv commented on pkg/raft/raft.go:
```diff
@@ -1744,7 +1744,23 @@ func (r *raft) Step(m pb.Message) error {
-			r.send(pb.Message{To: m.From, Term: m.Term, Type: voteRespMsgType(m.Type)})
+			var logHint entryID
+			if m.Type == pb.MsgVote {
+				logHint = lastID
```
Looking at this now, after https://github.com/cockroachdb/cockroach/pull/143713#discussion_r2059287244, I think this hint could be improved further. Using `candLastID`, we can run it through one round of `findConflictByTerm`, and attach the result of it.

This utilizes the fact that `findConflictByTerm` is symmetric, and e.g. `MsgApp` protocol uses it on both ends the same way.

pav-kv replied:
Worth leaving a TODO.

pav-kv replied:
> further improvement would involve attaching the candidate lastEntryID into MsgVote

This is already attached. It's the `candLastID` variable here. When requesting votes, the candidate attaches its last entry ID (so that the peers can decide whether to vote for it).

hakuuww commented on pkg/raft/raft_test.go:
```diff
@@ -2685,6 +2685,194 @@ func TestLeaderAppResp(t *testing.T) {
+				index: 0,
+			},
+			next:    11,
+			prState: tracker.StateProbe,
```
Yes, the diagrams just illustrates a possible follower log, with the specific last entryID that gets attached to the vote.

hakuuww committed:
raft: apis for leader storing hint information from MsgVoteResp

Epic: None
Release note: None

hakuuww committed:
raft: record and utilize hints from voting process after becoming leader

Upon leader receiving voter's votes from MsgVoteResp{reject = false}, use
the hint "voterHint" entryID to compute a best guess conflict index.

Record the computed best guess index into the election tracker.

After winning election and becoming leader,use the hints recorded during the
election phase to set follower progress to optimize for the first round of
MsgApp/replication.

If the follower voted for this new leader during election phase, then the
leader can either:
 - If we know that the voter's log matches ours, send the first round of
   MsgApp starting at the correct fork point and set the follower's state
   to StateReplicate.
 - If the voter's log doesn't match ours, send the first round of MsgApp at
   a best guess index.

Special Case:
If the computed hint index is below or equal to the candidate/new-leader's
compacted index, we would want to send a snapshot to that voter immediately
after becoming leader.
To allow this to happen, Progress.RecentActive is set to true if the voter voted
for the candidate during election. This is because RecentActive is used as a
blocker to decide whether to send a snapshot to a follower.
This should be safe as we already set Progress.RecentActive to true if the leader
receives a MsgAppResp. In this case, since we are pipelining probing logic into
MsgVoteResp, it makes intuitive sense to do this change.

A detailed unit test is added to test for different types of behaviors that
may happen. Including the cases where the voter needs a snapshot (based on its
provided hint).

See interaction test output changes to examine behavior changes.

For further optimizations:
The "term cache" can be utilized to handle more extreme divergent log tail cases.

Epic: none
Fixes: #143713
Release note (performance improvement): New raft leader replicates to followers faster after winning election.

hakuuww committed:
raft: modify snapshot_succeed_via_app_resp_behind.txt

Previously, n1(leader) does not send snapshot to the behind follower n3 due
 to vote pipelining modifying how the leader sees n3. This commit drops
n3's response to n1(candidate) which ultimately allows n1 to send snapshot
to n3 and test the original intent of the test.

Epic: none
Release note: None

hakuuww commented:
@pav-kv RDFR

hakuuww commented on pkg/raft/raft.go:
```diff
@@ -2368,8 +2407,48 @@ func leadSliceFromMsgApp(m *pb.Message) LeadSlice {
+		// index <= voterHint.index.
+		// Also, raftLog.term(index) <= voterHint.term, or the entry at index
+		// is missing.
+		index, term := r.raftLog.findConflictByTerm(voterHint.index, voterHint.term)
```
Subsequently, when the candidate sees the hint from the voter, it can rely on that fact that voterHint is a best guess hint, and it can go into findConflictByTerm without needing any other processing.

hakuuww commented on pkg/raft/raft.go:
```diff
@@ -1757,7 +1757,45 @@ func (r *raft) Step(m pb.Message) error {
 			// same in the case of regular votes, but different for pre-votes.
-			r.send(pb.Message{To: m.From, Term: m.Term, Type: voteRespMsgType(m.Type)})
+			var logHint entryID
+			if m.Type == pb.MsgVote {
+				// Return a hint to the candidate - a guess on the maximal (index, term)
+				// at which the logs match. Do this by searching through the voter's
+				// log for the maximum (index, term) pair with a term <= candLastID.term
+				// and an index <= the candLastID.index
+				//
+				// The candidate's log can shorter than the voter's. So we take the
+				// minimum of the two log's last index to pass into findConflictByTerm.
+				//
+				// If the resulting logHint.index and logHint.term are both non-zero,
+				// it indicates that the voter's log has an entry with
+				// a term <= candLastID.term and an index <= candLast.index.
+				//
+				// It is possible that one or both of the returning index and term values
+				// are 0. This is fine as the candidate would ignore the hint in such case.
+				//
+				// This can help skip all indexes in the voter's uncommitted log tail
+				// with terms greater than candLast.term.
+				//
+				// The use of findConflictByTerm here is very similar to the one in
+				// handleAppendEntries. See the other caller for findConflictByTerm
+				// (in stepLeader) for a much more detailed explanation of this mechanism.
+				minIdx := min(candLastID.index, lastID.index)
+				logHint.index, logHint.term = r.raftLog.findConflictByTerm(minIdx, candLastID.term)
+			}
```
Pending on: #145192

hakuuww commented on pkg/raft/raft.go:
```diff
@@ -1735,16 +1787,43 @@ func (r *raft) Step(m pb.Message) error {
+				//     logHInt.term <= candLastID.term
+				//
+				// See comments in findConflictByTerm for detailed explanation.
+				logHint.index, logHint.term = r.raftLog.findConflictByTerm(candLastID.index, candLastID.term)
```
> Can remove the call to minIdx := min(candLastID.index, lastID.index) after https://github.com/cockroachdb/cockroach/pull/145192 is done.

Done.

hakuuww commented on pkg/raft/raft.go:
```diff
@@ -1748,16 +1748,43 @@ func (r *raft) Step(m pb.Message) error {
+				// the message (it ignores all out of date messages).
+				// The term in the original message and current local term are the
+				// same in the case of regular votes, but different for pre-votes.
+				Term: m.Term,
```
Need to avoid polluting git blame here.

hakuuww replied:
Done. Leaving it at original location
